tornado/httpserver.py

```diff
@@ -179,12 +179,78 @@ def handle_stream(self, stream, address):
         conn.start_serving(self)
 
     def start_request(self, server_conn, request_conn):
-        return _ServerRequestAdapter(self, server_conn, request_conn)
+        if isinstance(self.request_callback, (tuple, list)):
```

Comment:
This can be done. In fact I removed the separate HTTPServerRouter class in the last versions of the commit to simplify the code.